### ISSUE TITLE
STATEMIND-1: UniswapV2 LP-Token price oracle can be manipulated

### ISSUE THREAD

**Issue by jpick713** [mysofinance/v2#364]
Description
The contract is using the following formula to calculate LP-Token price:

price = 2 * (sqrt(reserve0 * reserve1) * sqrt(price0 * price1)) / totalSupply

The method used to calculate LP-Token price holds well if the reserve pools stay balanced according to constant product formula: x * y = k. However UniswapV2 Pair contract allows to arbitrary transfer tokens to reserve without minting new LP-Tokens or adhering to contstant product formula via sync().

The attacker can increase the amount of tokens in one of the reserves by directly transferring the tokens to the UniswapV2 Pair contract and then executing the sync function, which will update the reserve balances.

This manipulation can only increase the amount of tokens in the reserves and consecutively the price of LP-Token, but not to decrease. So to attack the protocol, the attacker would need to use their own LP-Tokens as the collateral (not for borrowing LP-Tokens).

It should be noted that sending tokens to the UniswapV2 Pair this way won’t mint any new LP-shares and the attacker won’t be able to fully return their funds, however, skewing reserves in such way creates an arbitrage opportunity that the attacker can use to partially reclaim the funds sent to uniswap pool.

To summarize a potential attack scenario:

Mint or borrow LP-Tokens
Send large amount of TokenA or TokenB to UniswapV2 Pair
Call sync() on UniswapV2 Pair, imbalancing the reserves
Use inflated LP-Tokens to borrow assets
Partially return the sent tokens from UniswapV2 Pair by rebalancing reserves back from self-created arbitrage opportunity
The economic feasibility of this attack largely depends on size of reserves pre-attack and further complicated by need of having LP-Tokens to use in borrowing positions.

Recommendation
It is recommended not to use this oracle for pairs with low liquidity or ERC20 tokens with small amount of decimals.
